Implement a software fallback algorithm for core.bitop.bsf() and bsr()

tsbockman opened the pull request:
This pull request is intended to supersede [PR #1486](https://github.com/D-Programming-Language/druntime/pull/1486#issuecomment-199689712).

This fixes an issue identified by @denis-sh with the `ulong` versions that [I previously added](https://github.com/D-Programming-Language/druntime/pull/1452) to `core.bitop` for `bsf()` and `bsr()`: the `ulong` overloads were defined to forward to the `size_t` versions on 64-bit, but then `size_t` is just an alias for `ulong`, which would make the function infinitely recursive.

This was not a problem in testing, because on 64-bit the `ulong` function body is replaced by a compiler intrinsic on all current DMD platforms.

Still, this pull request solves this potential issue by adding a software fallback algorithm for bit scans - `softScan()` - just in case someone wants to use `bsf()` or `bsr()` on a platform where they are not intrinsics.

There is no longer an explicit `size_t` overload of `bsf()` or `bsr()`; instead there are separate `uint` and `ulong` overloads, each of which forwards to the overload which is expected to be an intrinsic on the current platform. The expected intrinsic, in turn, falls back to `softScan()` if needed.

denis-sh commented:
Yes, let's add more useless functions and increase code size with `softScan`. The more the better, isn't it?

tsbockman commented:
> Yes, let's add more useless functions and increase code size with `softScan`.

Why did you even complain and make your pull request in the first place, if you don't care about making `bsf()`/`bsr()` work without the intrinsics? Just to be pedantic?

Your PR doesn't actually *do* anything as far as I can tell.

denis-sh commented:
> Why did you even complain and make your pull request in the first place, if you don't care about making bsf()/bsr() work without the intrinsics? Just to be pedantic?

Two reasons:

1. There should be no useless code in the library.
2. There should be no wrong code in a library even if it's unused now (for unused code see 1).

denis-sh commented:
> Your PR doesn't actually do anything as far as I can tell.

My PR removes 64-bit `bsf`/`bsr` implementation in 64 bit mode, i.e. removes unused code.

tsbockman commented:
OK. I'll let the reviewers decide whether they want my fallback implementation or not.

CyberShadow commented:
> Yes, let's add more useless functions and increase code size with softScan. The more the better, isn't it?

Please be nice.

WalterBright commented on src/core/bitop.d:
```diff
@@ -142,6 +156,74 @@ unittest
     assert(test_ctfe == 63);
 }
 
+private alias softBsf(N) = softScan!(N, true);
```
Function is not documented.

WalterBright commented on src/core/bitop.d:
```diff
@@ -142,6 +156,74 @@ unittest
+        else
+            ret += inc!16;
+    }
+    else
```
else what values of N are acceptable? Clearly `ubyte` won't work, but there needs to be a test for what the code is designed to work with.

WalterBright replied:
Ah, I missed it must be `uint`. Perhaps add `else static if (is(N == uint))` and then an `else static assert(0);`

WalterBright commented on src/core/bitop.d:
```diff
@@ -142,6 +156,74 @@ unittest
+    if (!v)
+        return -1;
+
+    N x = void;
```
`= void` is not necessary, as standard dead assignment analysis will remove it.

WalterBright commented on src/core/bitop.d:
```diff
@@ -142,6 +156,74 @@ unittest
+        return -1;
+
+    N x = void;
+    int ret = void;
```
ditto

WalterBright commented on src/core/bitop.d:
```diff
@@ -142,6 +156,74 @@ unittest
+
+    return ret;
+}
+
```
Need unit tests. No reason not to get 100% line coverage.

tsbockman commented:
I added documentation and a fairly thorough `unittest` for `softScan`, as well as made the requested style tweaks.

@WalterBright, I believe I have addressed all of your comments, unless the `softBsf` and `softBsr` aliases really need to be documented separately from `softScan`? I can't think why they would though, since these are all private and there's nothing to say about the aliases that isn't already right there in the code.

ibuclaw commented on src/core/bitop.d:
```diff
@@ -142,6 +156,110 @@ unittest
+
+    // This is essentially an unrolled binary search:
+    enum mask(ulong lo) = forward? cast(N) lo : cast(N)~lo;
+    enum inc(int up) = forward? up : -up;
```
What is this, ruby? :-)

Though I'm not immediately sure of any coding convention imposed on condition expressions in druntime/phobos.

tsbockman replied:
> What is this, ruby? :-)

For what it's worth, I've never written so much as one line of ruby in my life. I do like the ternary operator, though.

schuetzm replied:
He means that you should insert a space before the `?`. (In Ruby, method names can end in `?`.)

tsbockman commented:
@ibuclaw D-rubyfied.

DmitryOlshansky commented:
LGTM Anything else @WalterBright?

DmitryOlshansky commented:
Auto-merge toggled on

tsbockman commented:
@DmitryOlshansky Since you merged this, #1486 should be closed now?

9il commented:
This should be reverted. Please see #1694

tsbockman commented:
> This should be reverted.

How about we *fix* it instead?

9il commented:
I think it should be reverted first

tsbockman commented:
Have you made any effort to understand why the intrinsics aren't working? Where's your test code? What compiler(s) are you having problems with?

9il commented:
I have - http://forum.dlang.org/post/ajifmjpyrijddhiblaky@forum.dlang.org. Why do you think they work (i mean BSR/BSF instructions for x86)? Where is your test code?

tsbockman commented:
> I think it should be reverted first

That would be a breaking change. The practical motivation for this PR was to get `bsf()` and `bsr()` working at CTFE.

> Why do you think they work (i mean BSR/BSF instructions for x86)?

Because I tested them when I submitted this PR. I probably only tested the intrinsic aspect with DMD though. (LDC had it's own druntime implementation at the time, with many of the differences relating to intrinsic handling. I figured they would figure out how to integrate it, as they had obviously done for past intrinsic work.)

> Where is your test code?

Gone. I didn't keep it once this got pulled.

> I have - http://forum.dlang.org/post/ajifmjpyrijddhiblaky@forum.dlang.org.

Thanks. I'll take a look.

tsbockman commented:
> I have - http://forum.dlang.org/post/ajifmjpyrijddhiblaky@forum.dlang.org.

That's not a valid test. Intrinsics are matched to fully qualified names; once you copy-paste the code to another module it stops matching.

If you actually `import core.bitop;` as intended, it works fine on DMD and GDC. LDC additionally requires `-enable-cross-module-inlining`, as noted by [Johan Engelen](http://forum.dlang.org/post/gbqfcshdayxgcxcmrgkx@forum.dlang.org).

9il commented:
No, LDC and GDC have not recent Druntime. Please provide poof for recent DMD

tsbockman commented:
> Please provide poof for recent DMD

It's currently working correctly for `bsr(ulong)` on 64-bit:
```
0000000000427cc8 <_Dmain>:
_Dmain():
/home/tsbockman/Code/d-scratch/source/app.d:5
module app;

import core.bitop, core.time;

int main() {
  427cc8:	55                   	push   rbp
  427cc9:	48 8b ec             	mov    rbp,rsp
  427ccc:	48 83 ec 10          	sub    rsp,0x10
/home/tsbockman/Code/d-scratch/source/app.d:6
    const rtval = MonoTime.currTime.ticks();
  427cd0:	e8 87 05 00 00       	call   42825c <_D4core4time42__T12MonoTimeImplVE4core4time9ClockTypei0Z12MonoTimeImpl8currTimeFNbNdNiNeZS4core4time42__T12MonoTimeImplVE4core4time9ClockTypei0Z12MonoTimeImpl>
  427cd5:	48 89 45 f8          	mov    QWORD PTR [rbp-0x8],rax
  427cd9:	48 8d 7d f8          	lea    rdi,[rbp-0x8]
  427cdd:	e8 ca 06 00 00       	call   4283ac <_D4core4time42__T12MonoTimeImplVE4core4time9ClockTypei0Z12MonoTimeImpl5ticksMxFNaNbNdNiNfZl>
  427ce2:	48 89 45 f0          	mov    QWORD PTR [rbp-0x10],rax
/home/tsbockman/Code/d-scratch/source/app.d:7
    return bsr(rtval);
  427ce6:	48 0f bd 45 f0       	bsr    rax,QWORD PTR [rbp-0x10]
/home/tsbockman/Code/d-scratch/source/app.d:8
}
  427ceb:	c9                   	leave  
  427cec:	c3                   	ret    
  427ced:	0f 1f 00             	nop    DWORD PTR [rax]
```
But **NOT** for `bsr(uint)` on 64-bit for some reason:
```
0000000000427d08 <_Dmain>:
_Dmain():
/home/tsbockman/Code/d-scratch/source/app.d:5
module app;

import core.bitop, core.time;

int main() {
  427d08:	55                   	push   rbp
  427d09:	48 8b ec             	mov    rbp,rsp
  427d0c:	48 83 ec 10          	sub    rsp,0x10
/home/tsbockman/Code/d-scratch/source/app.d:6
    const rtval = cast(uint) MonoTime.currTime.ticks();
  427d10:	e8 53 06 00 00       	call   428368 <_D4core4time42__T12MonoTimeImplVE4core4time9ClockTypei0Z12MonoTimeImpl8currTimeFNbNdNiNeZS4core4time42__T12MonoTimeImplVE4core4time9ClockTypei0Z12MonoTimeImpl>
  427d15:	48 89 45 f8          	mov    QWORD PTR [rbp-0x8],rax
  427d19:	48 8d 7d f8          	lea    rdi,[rbp-0x8]
  427d1d:	e8 96 07 00 00       	call   4284b8 <_D4core4time42__T12MonoTimeImplVE4core4time9ClockTypei0Z12MonoTimeImpl5ticksMxFNaNbNdNiNfZl>
  427d22:	89 45 f0             	mov    DWORD PTR [rbp-0x10],eax
/home/tsbockman/Code/d-scratch/source/app.d:7
    return bsr(rtval);
  427d25:	8b 7d f0             	mov    edi,DWORD PTR [rbp-0x10]
  427d28:	e8 53 01 00 00       	call   427e80 <_D4core5bitop3bsrFNaNbNiNfkZi>
/home/tsbockman/Code/d-scratch/source/app.d:8
}
  427d2d:	c9                   	leave  
  427d2e:	c3                   	ret    
  427d2f:	90                   	nop
```
```
0000000000427e80 <_D4core5bitop3bsrFNaNbNiNfkZi>:
_D4core5bitop3bsrFNaNbNiNfkZi():
  427e80:	55                   	push   rbp
  427e81:	48 8b ec             	mov    rbp,rsp
  427e84:	48 83 ec 10          	sub    rsp,0x10
  427e88:	89 7d f8             	mov    DWORD PTR [rbp-0x8],edi
  427e8b:	8b 7d f8             	mov    edi,DWORD PTR [rbp-0x8]
  427e8e:	e8 05 00 00 00       	call   427e98 <_D4core5bitop19__T8softScanTmVbi0Z8softScanFNaNbNiNfmZi>
  427e93:	48 8b e5             	mov    rsp,rbp
  427e96:	5d                   	pop    rbp
  427e97:	c3                   	ret    
```
I will have to investigate further and see if I can figure out why it's taking the wrong `static if` branch for `uint`.

tsbockman commented:
It seems that DMD's intrinsic detection fails randomly if there is a function body available: https://issues.dlang.org/show_bug.cgi?id=16743

It looks like a compiler issue, not a druntime one; GDC and LDC are unaffected.

tsbockman commented:
It turns out that intrinsics and inlining don't mix. A fix is in the works:

[DMD PR #6278](https://github.com/dlang/dmd/pull/6278)
[druntime PR #1695](https://github.com/dlang/druntime/pull/1695)
